### PR TITLE
Online accountancy

### PR DESCRIPTION
**Pull request by MJHutchinson**

### CODE REVIEW

File: src/privacy_accounting/analysis/pld_accountant/pld_accountant.py [hunk 1]
```diff
@@ -444,14 +485,81 @@ def compute_privacy_loss_from_ledger(ledger, target_eps=None, target_delta=None,
...
     if adjacency_definition is 'add_remove':
         if target_eps is not None:
-            return get_delta_add_remove(effective_z_t, q_t, target_eps=target_eps, nx=nx, L=L)
+            privacy_bound, _ = get_delta_add_remove(effective_z_t, q_t, target_eps=target_eps, nx=nx, L=L)
+            return privacy_bound
+        else:
+            privacy_bound, _ = get_eps_add_remove(effective_z_t, q_t, target_delta=target_delta, nx=nx, L=L)
+            return privacy_bound
+
+    if adjacency_definition is 'add_remove':
```
Comment (MrinankSharma) [inline src/privacy_accounting/analysis/pld_accountant/pld_accountant.py]: typo - should be substitution

Comment (MrinankSharma) [reply src/privacy_accounting/analysis/pld_accountant/pld_accountant.py]: i also have a slightly preference for an else if

File: src/privacy_accounting/analysis/pld_accountant/pld_accountant.py [hunk 2]
```diff
@@ -444,14 +485,81 @@ def compute_privacy_loss_from_ledger(ledger, target_eps=None, target_delta=None,
 
     if adjacency_definition is 'add_remove':
         if target_eps is not None:
-            return get_delta_add_remove(effective_z_t, q_t, target_eps=target_eps, nx=nx, L=L)
+            privacy_bound, _ = get_delta_add_remove(effective_z_t, q_t, target_eps=target_eps, nx=nx, L=L)
+            return privacy_bound
+        else:
```
Comment (MrinankSharma) [inline src/privacy_accounting/analysis/pld_accountant/pld_accountant.py]: could you check target delta and return an error if it is blank ?

File: src/privacy_accounting/analysis/pld_accountant/pld_accountant.py [hunk 3]
```diff
@@ -444,14 +485,81 @@ def compute_privacy_loss_from_ledger(ledger, target_eps=None, target_delta=None,
...
+            privacy_bound, _ = get_eps_add_remove(effective_z_t, q_t, target_delta=target_delta, nx=nx, L=L)
+            return privacy_bound
+
+    if adjacency_definition is 'add_remove':
+        if target_eps is not None:
+            privacy_bound, _ = get_delta_substitution(effective_z_t, q_t, target_eps=target_eps, nx=nx, L=L)
+            return privacy_bound
+        else:
```
Comment (MrinankSharma) [inline src/privacy_accounting/analysis/pld_accountant/pld_accountant.py]: same here

File: src/privacy_accounting/analysis/pld_accountant/pld_accountant.py (hunk 2 above)
Comment (MrinankSharma) [reply src/privacy_accounting/analysis/pld_accountant/pld_accountant.py]: actually ignore - i didnt see above check

File: src/privacy_accounting/analysis/pld_accountant/pld_accountant.py (hunk 3 above)
Comment (MrinankSharma) [reply src/privacy_accounting/analysis/pld_accountant/pld_accountant.py]: ignore - didnt see above check

File: src/privacy_accounting/analysis/pld_accountant/pld_accountant.py [hunk 4]
```diff
@@ -444,14 +485,81 @@ def compute_privacy_loss_from_ledger(ledger, target_eps=None, target_delta=None,
...
+            privacy_bound, F_prod = get_delta_add_remove(effective_z_t, q_t, target_eps=target_eps, nx=nx, L=L,
+                                                         F_prod=F_prod)
+            return privacy_bound, F_prod
         else:
-            return get_eps_add_remove(effective_z_t, q_t, target_delta=target_delta, nx=nx, L=L)
+            privacy_bound, F_prod = get_eps_add_remove(effective_z_t, q_t, target_delta=target_delta, nx=nx, L=L,
+                                                       F_prod=F_prod)
+            return privacy_bound, F_prod
 
     if adjacency_definition is 'add_remove':
```
Comment (MrinankSharma) [inline src/privacy_accounting/analysis/pld_accountant/pld_accountant.py]: should be substitution

File: src/privacy_accounting/optimizer/dp_optimizer.py [hunk 5]
```diff
@@ -4,19 +4,19 @@
...
 from src.privacy_accounting.dp_query import GaussianDPQuery
-from . import WrapperOptimizer
+from .wrapper_optimizer import WrapperOptimizer
 
 
 class DPOptimizer(WrapperOptimizer):
 
     def __init__(self,
-                 optimizer,
+                 optimiser,
```
Comment (MrinankSharma) [inline src/privacy_accounting/optimizer/dp_optimizer.py]: omg all of these z have become an s!! pls no !! we are using z everywhere atm

File: src/privacy_accounting/analysis/moment_accountant/moment_accountant.py [hunk 6]
```diff
@@ -97,3 +97,14 @@ def compute_privacy_loss_from_ledger(ledger, target_delta=None, target_eps=None,
 
     total_log_moments = compute_log_moments_from_ledger(ledger, max_lambda)
     return get_privacy_spent(max_lambda, total_log_moments, target_eps, target_delta)
+
+
+def compute_online_privacy_from_ledger(ledger, log_moments, target_delta=None, target_eps=None, max_lambda=32):
```
Comment (MrinankSharma) [inline src/privacy_accounting/analysis/moment_accountant/moment_accountant.py]: could you specify that this is the incremental ledger

File: src/privacy_accounting/analysis/pld_accountant/pld_accountant.py [hunk 7]
```diff
@@ -444,14 +485,81 @@ def compute_privacy_loss_from_ledger(ledger, target_eps=None, target_delta=None,
...
+
+    raise ValueError('adjacency_definition must be one of "substitution" or "add_remove".')
+
+
+def compute_online_privacy_from_ledger(ledger, F_prod,
+                                       target_delta=None, target_eps=None,
+                                       adjacency_definition='add_remove', nx=1E6, L=20.0):
+    """ Compute new PLD privacy in an online fashion, to speed up computation.
+
+    :param ledger: The ledger of queries to compute for
```
Comment (MrinankSharma) [inline src/privacy_accounting/analysis/pld_accountant/pld_accountant.py]: could you state that this is the incremental ledger

File: src/privacy_accounting/optimizer/dp_optimizer.py (hunk 5 above)
Comment (MJHutchinson) [reply src/privacy_accounting/optimizer/dp_optimizer.py]: Yeah I think this is from the silly number of branches open

Comment (MJHutchinson) [reply src/privacy_accounting/optimizer/dp_optimizer.py]: Have fixed project wide I hope

File: src/utils/numpy_nest_utils.py [hunk 8]
```diff
@@ -0,0 +1,28 @@
...
+            items.append((new_key, v))
+    return dict(items)
+
+
+def structured_ndarrays_to_lists(d):
+    ret = {}
+    for k, v in d.items():
+        if isinstance(v, collections.MutableMapping):
+            ret[k] = structured_ndarrays_to_lists(v)
+        elif isinstance(v, np.ndarray):
```
Comment (MrinankSharma) [inline src/utils/numpy_nest_utils.py]: what if v is a scalar? or do you think we'd always use lists of length one?

**Pull request merged**